sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -850,6 +851,15 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       Row(null, null, 6, "F") :: Nil)
   }
 
+  test("SPARK-11111 null-safe join should not use cartesian product") {
+    val df = sql("select count(*) from testData a join testData b on (a.key <=> b.key)")
+    val cp = df.queryExecution.executedPlan.collect {
+      case cp: CartesianProduct => cp
```

Comment:
ah one thing about this test case: it might be better to assert the proper non-cartesian join operator is in place, rather than testing the cartesian product is not in place. The reason I'm bringing this up is that I'm sure as soon as we do local iterators (or other operator changes), this test will be a no-op (continue passing even if the implementation is wrong).